### ISSUE TITLE
Refactoring Contact Field backend

### ISSUE THREAD

**Issue by dodobas** [rapidpro/rapidpro#706]
User defined contact fields are currently stored in a EVA relation `values_value` which is also reused by flow run results. The table has grown in size and we are experiencing problems managing contact field data.

We are considering these options:

### Continue using the current EVA model
  * move contact fields to a new relation
  * optimize SQL statements
    * for example instead of NOT IN (), use NOT EXISTS or LEFT JOIN
      * this is used in user searched like `joined=""`
    * NOT IN can not be optimized by PostgreSQL, NOT EXISTS or LEFT JOIN, which can be optimized as an 'anti join'
    * https://explainextended.com/2009/09/16/not-in-vs-not-exists-vs-left-join-is-null-postgresql/
  * minimal updates to the source code

### Move from the EVA to a 'document' field
  * we would use JSONB/HSTORE to store contact fields
  * potentially faster CRUD, because contact field data is 'closer' to the Contact
  * fast 'does it contain' key or 'does it contain' specific key=value, operators (with indices), but slow with <=, >= and limited support of data types
    * index does not collect statistics about data, causes problems for the planner
    * https://blog.heapanalytics.com/when-to-avoid-jsonb-in-a-postgresql-schema/
  * we could use functional indexes to prepare data for querying and 'cast' data to a desired type, but we would need to create an index management system
  * bigger source code update than the first option

### Move from EVA model to multiple columns
  * in a nutshell, every contact field is a normal attribute in the database
  * adding new attributes is cheap (in PostgreSQL this is just a system metadata update), NULL value storage is almost 'free'
  * we could partition contact fields data per `Org` so that every Org has it's own storage (table) for contact fields
  * biggest source code updates
    * partitioning system should be transparent to Django

### Move to ElasticSearch or some other 3rd party system
  * introducing new moving pieces to the platform is probably not the best option
  * we have limited practical knowledge
  * do we really need features available on a new platform
  * potential problems with data sync

**Comment by rowanseymour** [comment]
I'm in favour of trying the optimizations you describe in (1) and also not making any fundamental changes until we get rid of flow results in values. Do you have a test database that you can use to try testing out those optimizations?

**Comment by dodobas** [comment]
Although someone could probably optimize these queries, I got some interesting results. Using `INNER JOIN` in place of `id IN ()`,it is fast for a small result set (0.2% records in the table). However, when applying the same idea to a big result set (78% records in the table) results swing in the other direction.

# Search results for 1000000 contacts (org_id=1)

* data generated with `test_db generate --orgs 1 --contacts 1000000`

## Small result set - 2297 contacts

* `(gender = "M" and district="Faskari") OR (gender="F" and district="Zuru")`

### Django ORM generated query, execution time **~1600ms**
    
    SELECT COUNT(*) AS "__count"
    FROM "contacts_contact"
    INNER JOIN "contacts_contactgroup_contacts"
        ON ("contacts_contact"."id" = "contacts_contactgroup_contacts"."contact_id")
    WHERE ("contacts_contactgroup_contacts"."contactgroup_id" = 1 AND "contacts_contact"."org_id" = 1 AND
        (("contacts_contact"."id" IN (SELECT U0."contact_id" AS Col1
                                        FROM "values_value" U0
                                        WHERE (U0."contact_field_id" || '|' ||
                                                UPPER(SUBSTRING(U0."string_value", 1, 32))) = '1|M')
            AND "contacts_contact"."id" IN (SELECT V0."contact_id" AS Col1
                                            FROM "values_value" V0
                                            WHERE (V0."contact_field_id" = 5 AND
                                                    V0."location_value_id" IN (SELECT U0."id" AS Col1
                                                                            FROM "locations_adminboundary" U0
                                                                            WHERE (U0."level" = 2 AND
                                                                                    UPPER(U0."name" :: TEXT) =
                                                                                    UPPER('Faskari')))))
            ) OR
            ("contacts_contact"."id" IN (SELECT U0."contact_id" AS Col1
                                        FROM "values_value" U0
                                        WHERE (U0."contact_field_id" || '|' ||
                                                UPPER(SUBSTRING(U0."string_value", 1, 32))) = '1|F')
            AND "contacts_contact"."id" IN (SELECT V0."contact_id" AS Col1
                                            FROM "values_value" V0
                                            WHERE (V0."contact_field_id" = 5 AND
                                                    V0."location_value_id" IN (SELECT U0."id" AS Col1
                                                                            FROM "locations_adminboundary" U0
                                                                            WHERE (U0."level" = 2 AND
                                                                                    UPPER(U0."name" :: TEXT) =
                                                                                    UPPER('Zuru')))))
            )
        ) AND
        "contacts_contact"."is_test" = FALSE);

* https://explain.depesz.com/s/3aJ1

### Handwritten optimization, execution time **~120ms**
    
    SELECT count(*) AS "__count"
    FROM (
       SELECT *
       FROM "contacts_contact"
         INNER JOIN "contacts_contactgroup_contacts"
           ON "contacts_contact"."id" = "contacts_contactgroup_contacts"."contact_id" AND
              "contacts_contactgroup_contacts"."contactgroup_id" = 1 AND "contacts_contact"."org_id" = 1 AND
              "contacts_contact"."is_test" = FALSE
         INNER JOIN values_value v ON contacts_contact.id = v.contact_id AND
                                      v."contact_field_id" || '|' || UPPER(SUBSTRING(v."string_value", 1, 32)) = '1|M'
         INNER JOIN values_value v1 ON contacts_contact.id = v1.contact_id AND
                                       (v1."contact_field_id" = 5 AND v1."location_value_id" IN (SELECT U0."id" AS Col1 FROM "locations_adminboundary" U0
                                                                                                 WHERE (U0."level" = 2 AND UPPER(U0."name" :: TEXT) = UPPER('Faskari'))))

       UNION

       SELECT *
       FROM "contacts_contact"
         INNER JOIN "contacts_contactgroup_contacts"
           ON "contacts_contact"."id" = "contacts_contactgroup_contacts"."contact_id" AND
              "contacts_contactgroup_contacts"."contactgroup_id" = 1 AND "contacts_contact"."org_id" = 1 AND
              "contacts_contact"."is_test" = FALSE
         INNER JOIN values_value v ON contacts_contact.id = v.contact_id AND
                                      v."contact_field_id" || '|' || UPPER(SUBSTRING(v."string_value", 1, 32)) = '1|F'
         INNER JOIN values_value v1 ON contacts_contact.id = v1.contact_id AND
                                       (v1."contact_field_id" = 5 AND v1."location_value_id" = (SELECT U0."id" AS Col1 FROM "locations_adminboundary" U0
                                                                                                WHERE (U0."level" = 2 AND UPPER(U0."name" :: TEXT) = UPPER('Zuru'))))
     ) AS new_table;

* https://explain.depesz.com/s/jg6

## Big result set - 780387 contacts

* `(gender = "M") OR (gender="F")`

### Django ORM generated query, execution time **~1700ms**

    SELECT COUNT(*) AS "__count"
    FROM "contacts_contact"
        INNER JOIN "contacts_contactgroup_contacts"
        ON ("contacts_contact"."id" = "contacts_contactgroup_contacts"."contact_id")
        WHERE ("contacts_contactgroup_contacts"."contactgroup_id" = 1 AND "contacts_contact"."org_id" = 1 AND
        (("contacts_contact"."id" IN (SELECT U0."contact_id" AS Col1
                                        FROM "values_value" U0
                                        WHERE (U0."contact_field_id" || '|' ||
                                                UPPER(SUBSTRING(U0."string_value", 1, 32))) = '1|M')
            ) OR
            ("contacts_contact"."id" IN (SELECT U0."contact_id" AS Col1
                                        FROM "values_value" U0
                                        WHERE (U0."contact_field_id" || '|' ||
                                                UPPER(SUBSTRING(U0."string_value", 1, 32))) = '1|F')
            )
        ) AND
        "contacts_contact"."is_test" = FALSE);

* https://explain.depesz.com/s/JVAD

### Handwritten (same) optimization, execution time **~7600ms**

    SELECT count(*) AS "__count"
    FROM (
        SELECT *
        FROM "contacts_contact"
            INNER JOIN "contacts_contactgroup_contacts"
            ON "contacts_contact"."id" = "contacts_contactgroup_contacts"."contact_id" AND
                "contacts_contactgroup_contacts"."contactgroup_id" = 1 AND "contacts_contact"."org_id" = 1 AND
                "contacts_contact"."is_test" = FALSE
            INNER JOIN values_value v ON contacts_contact.id = v.contact_id AND
                                        v."contact_field_id" || '|' || UPPER(SUBSTRING(v."string_value", 1, 32)) = '1|M'
        UNION

        SELECT *
        FROM "contacts_contact"
            INNER JOIN "contacts_contactgroup_contacts"
            ON "contacts_contact"."id" = "contacts_contactgroup_contacts"."contact_id" AND
                "contacts_contactgroup_contacts"."contactgroup_id" = 1 AND "contacts_contact"."org_id" = 1 AND
                "contacts_contact"."is_test" = FALSE
            INNER JOIN values_value v ON contacts_contact.id = v.contact_id AND
                                        v."contact_field_id" || '|' || UPPER(SUBSTRING(v."string_value", 1, 32)) = '1|F'
        ) AS new_table;

* https://explain.depesz.com/s/QV9n
  * because of the UNION data is sorted which gets swapped to disk (increasing work_mem would fix it), but even if we 'forget' about sort/swap operation, the query still takes a long time

**Comment by nicpottier** [comment]
Moved to JSON field with Elastic searching.
